internal/decryption/output.go

```diff
@@ -21,39 +21,87 @@ import (
 	"go.mongodb.org/mongo-driver/bson"
 )
 
-type AuditLogOutput struct {
-	Warningf  func(lineNb int, format string, a ...interface{}) error
-	Error     func(lineNb int, err error) error
-	Errorf    func(lineNb int, format string, a ...interface{}) error
-	LogRecord func(lineNb int, logRecord interface{}) error
-}
-
-func buildOutput(out io.Writer) AuditLogOutput {
-	newLine := []byte{'\n'}
-	writeLine := func(value []byte) error {
-		if _, err := out.Write(value); err != nil {
-			return err
-		}
-		_, err := out.Write(newLine)
+type AuditLogOutput interface {
+	Warningf(lineNb int, format string, a ...interface{}) error
+	Error(lineNb int, err error) error
+	Errorf(lineNb int, format string, a ...interface{}) error
+	LogRecord(lineNb int, logRecord interface{}) error
+}
+
+type auditLogOutputImpl struct {
+	out io.Writer
+}
+
+type AuditLogErrorLevel string
+
+const (
+	AuditLogErrorLevelError   AuditLogErrorLevel = "decryptionLogError"
+	AuditLogErrorLevelWarning AuditLogErrorLevel = "decryptionLogWarning"
+)
+
+type AuditLogError struct {
+	Level AuditLogErrorLevel
+	Line  int
+	Err   error
+}
+
+func (e AuditLogError) MarshalBSON() ([]byte, error) {
+	doc := bson.D{
+		bson.E{Key: "atype", Value: string(e.Level)},
+		bson.E{Key: "line", Value: e.Line},
+		bson.E{Key: "error", Value: e.Err.Error()},
+	}
+	return bson.Marshal(doc)
+}
+
+func (e AuditLogError) Error() string {
+	return e.Err.Error()
+}
+
+func NewAuditLogOutput(out io.Writer) AuditLogOutput {
+	return &auditLogOutputImpl{
+		out: out,
+	}
+}
```

Comment:
[future improvement] `auditLogOutputImpl` has potential to grow in customisations, consider making this `new` functional 
```suggestion
func NewAuditLogOutput(opts ...func(AuditLogOutput)) AuditLogOutput {
	a := &auditLogOutputImpl{
		out: out,
	}
	for _, o := range opts {
		o(a)
	}
	return a	
}
```
see https://github.com/mongodb/go-client-mongodb-atlas/blob/116acf986f4738d6a54e4ebaca13071ac3f86326/mongodbatlas/mongodbatlas.go#L307-L316